packages/portfolio-deploy/src/contract-control.js

```diff
@@ -0,0 +1,308 @@
+ */
+
+import { makeTracer } from '@agoric/internal/src/debug.js';
+import { Fail, q } from '@endo/errors';
+import { E, passStyleOf } from '@endo/far';
+import { M, mustMatch, objectMap } from '@endo/patterns';
+
+const trace = makeTracer('CCtrl');
+
+/**
+ * @import {Zone} from '@agoric/base-zone';
+ * @import {ContractStartFunction} from '@agoric/zoe/src/zoeService/utils';
+ * @import {IssuerKeywordRecord} from '@agoric/zoe';
+ * @import {Remote} from '@agoric/internal';
+ * @import {Board, NameHubKit} from '@agoric/vats';
+ * @import {CopyRecord} from '@endo/pass-style';
+ */
+
+const IssuerShape = M.remotable('Issuer');
+const InstallationShape = M.remotable('Installation');
+const KitShape = M.splitRecord(
+  {
+    instance: M.remotable('Instance'),
+    publicFacet: M.remotable('publicFacet'),
+    adminFacet: M.remotable('adminFacet'),
+    privateArgs: M.record(),
+    label: M.string(),
+  },
+  {
+    creatorFacet: M.or(M.remotable('creatorFacet'), undefined),
+    creatorInvitation: M.or(M.remotable('creatorFacet'), undefined),
+  },
+  {},
+);
+const BundleIdShape = M.string();
+const VatUpgradeResultsShape = M.splitRecord({ incarnationNumber: M.number() });
+
+// XXX work around exo interface guard conflict in the bootstrap vat
+const iface0 = {
+  install: M.string(),
+  start: M.splitRecord(
+    { installation: InstallationShape },
+    {
+      issuers: M.recordOf(M.string(), IssuerShape),
+      dataPrivateArgs: M.record(),
+    },
+  ),
+  installAndStart: M.splitRecord(
+    { bundleId: BundleIdShape },
+    {
+      issuers: M.recordOf(M.string(), IssuerShape),
+      dataPrivateArgs: M.record(),
+    },
+  ),
+  upgrade: BundleIdShape,
+  terminate: M.splitRecord(
+    {},
+    { message: M.string(), target: M.string(), revoke: M.boolean() },
+  ),
+  pruneChainStorage: M.recordOf(M.string(), M.arrayOf(M.string())),
+};
+harden(iface0);
+
+const iface = M.interface('ContractControl', {
+  install: M.callWhen(iface0.install).returns({
+    boardId: M.string(),
+    installation: InstallationShape,
+  }),
+  start: M.callWhen(iface0.start).returns(KitShape),
+  installAndStart: M.callWhen(iface0.installAndStart).returns(KitShape),
+  upgrade: M.callWhen().returns(VatUpgradeResultsShape),
+  terminate: M.callWhen(iface0.terminate).returns(),
+  pruneChainStorage: M.callWhen(iface0.pruneChainStorage).returns(M.number()),
+});
+harden(iface);
+
+/**
+ * @template {ContractStartFunction} SF
+ * @typedef {StartedInstanceKit<SF> & {
+ *   label: string;
+ *   privateArgs: Parameters<SF>[1];
+ * }} UpgradeKit
+ */
+
+/**
+ * @param {Zone} zone
+ * @param {{
+ *   agoricNamesAdmin: ERef<NameHubKit['nameAdmin']>,
+ *   board: ERef<Board>,
+ *   zoe: ERef<ZoeService>,
+ *   startUpgradable: ERef<StartUpgradable>,
+ * }} svcs
+ */
+export const prepareContractControl = (zone, svcs) => {
+  return zone.exoClass(
+    'ContractControl',
+    // iface,
+    undefined, // TODO: debug RangeError#13: Expected undefined is same as Interface
```

Comment:
I thought we had fixed Exo stuff such that we could run inside an old `VatData` environment. This basically should be a case of new code able to handle old JS hosts @erights

Reply:
Using a heap zone allowed me to revert this.